src/Symfony/Component/Messenger/MessageBus.php

```diff
@@ -11,28 +11,39 @@
 
 namespace Symfony\Component\Messenger;
 
-use Symfony\Component\Messenger\Exception\InvalidArgumentException;
 use Symfony\Component\Messenger\Middleware\MiddlewareInterface;
 
 /**
  * @author Samuel Roze <samuel.roze@gmail.com>
  * @author Matthias Noback <matthiasnoback@gmail.com>
+ * @author Nicolas Grekas <p@tchwork.com>
  */
 class MessageBus implements MessageBusInterface
 {
-    private $middlewareHandlers;
-
-    /**
-     * @var MiddlewareInterface[]|null
-     */
-    private $indexedMiddlewareHandlers;
+    private $middlewareAggregate;
 
     /**
      * @param MiddlewareInterface[]|iterable $middlewareHandlers
      */
     public function __construct(iterable $middlewareHandlers = array())
     {
-        $this->middlewareHandlers = $middlewareHandlers;
+        if ($middlewareHandlers instanceof \IteratorAggregate) {
+            $this->middlewareAggregate = $middlewareHandlers;
+        } elseif (\is_array($middlewareHandlers)) {
+            $this->middlewareAggregate = new \ArrayObject($middlewareHandlers);
+        } else {
+            $this->middlewareAggregate = new class() {
+                public $aggregate;
+                public $iterator;
+
+                public function getIterator()
+                {
+                    return $this->aggregate = new \ArrayObject(iterator_to_array($this->iterator, false));
+                }
+            };
+            $this->middlewareAggregate->aggregate = &$this->middlewareAggregate;
```

Comment:
Can you move that to be in the `constructor` of your `class` instead?

Reply:
The class is internal, the public properties are not accessible. Adding a constructor would just add unneeded boilerplate. Better not to me.